File: mobile/test/ride_together_test.dart
import 'package:aetherride_mobile/data/community/ride_together_cloud.dart';
import 'package:aetherride_mobile/domain/community/ride_group.dart';
import 'package:aetherride_mobile/domain/community/ride_group_policy.dart';
import 'package:aetherride_mobile/domain/community/ride_together.dart';
import 'package:flutter_test/flutter_test.dart';

RideGroup _g({
  required String id,
  required String route,
  required DateTime created,
  bool onServer = true,
}) {
  return RideGroup(
    id: id,
    hostUserId: 'h',
    savedRouteId: route,
    title: id,
    startWindowStart: created,
    startWindowEnd: created.add(const Duration(hours: 8)),
    joinCode: 'ABCDEF',
    status: RideGroupStatus.riding,
    livePinsAllowed: true,
    createdAt: created,
    onServer: onServer,
  );
}

void main() {
  test('Session-Route und Label', () {
    expect(RideTogetherPolicy.isSessionRouteId('freeride'), isTrue);
    expect(RideTogetherPolicy.isFreerideRide(null), isTrue);
    expect(RideTogetherPolicy.canJoinSessionByCode('freeride'), isTrue);
    expect(RideGroupPolicy.normalizeJoinCode('ab c2-d3'), 'ABC2D3');
    expect(RideTogetherPolicy.sanitizeLabel('  Luka   Flow  '), 'Luka Flow');
    expect(RideTogetherPolicy.sanitizeLabel('x' * 40).length, 24);
    expect(RideTogetherPolicy.matePair('b', 'a')?.lo, 'a');
    expect(RideTogetherPolicy.matePair('a', 'a'), isNull);
    expect(RideTogetherPolicy.canAddMember(19), isTrue);
    expect(RideTogetherPolicy.canAddMember(20), isFalse);
    expect(RideTogetherPolicy.sessionClosesAfterLeave(1), isFalse);
    expect(RideTogetherPolicy.sessionClosesAfterLeave(0), isTrue);
    expect(RideTogetherPolicy.stopLookClosesSession(), isFalse);
    expect(RideTogetherPolicy.stopLookClosesSoloSession(), isTrue);
    expect(
      RideTogetherPolicy.pickRequestSession(fromCount: 1, toCount: 1),
      'from',
    );
    expect(
      RideTogetherPolicy.pickRequestSession(fromCount: 8, toCount: 1),
      'from',
    );
    expect(
      RideTogetherPolicy.pickRequestSession(fromCount: 1, toCount: 8),
      'to',
    );
    expect(
      RideTogetherPolicy.pickRequestSession(fromCount: 5, toCount: 12),
      'none',
    );
  });

  test('Zelle: neben dir, nicht die Stadt', () {
    expect(
      RideTogetherPolicy.bucket(
        selfLat: 49.4094,
        selfLng: 8.6948,
        otherLat: 49.4094,
        otherLng: 8.6948,
      ),
      'beside',
    );
    expect(
      RideTogetherPolicy.bucket(
        selfLat: 49.4094,
        selfLng: 8.6948,
        otherLat: 49.5,
        otherLng: 8.8,
      ),
      isNull,
    );
  });

  test('pickGroupForRide: Freeride nur Session mit zwei Leuten', () {
    final session = _g(
      id: 's1',
      route: RideTogetherPolicy.routeId,
      created: DateTime.utc(2026, 8, 19, 10),
    );
    final planned = _g(
      id: 'p1',
      route: 'r-bodensee-road',
      created: DateTime.utc(2026, 8, 19, 9),
    );
    expect(
      RideTogetherPolicy.pickGroupForRide(
        rideRouteId: null,
        groups: [session, planned],
        memberCounts: {'s1': 1, 'p1': 2},
      ),
      isNull,
    );
    expect(
      RideTogetherPolicy.pickGroupForRide(
        rideRouteId: null,
        groups: [session, planned],
        memberCounts: {'s1': 2, 'p1': 2},
      )?.id,
      's1',
    );
    expect(
      RideTogetherPolicy.pickGroupForRide(
        rideRouteId: null,
        groups: [session],
        memberCounts: {'s1': 20},
      )?.id,
      's1',
    );
    expect(
      RideTogetherPolicy.pickGroupForRide(
        rideRouteId: 'r-bodensee-road',
        groups: [session, planned],
        memberCounts: {'s1': 2, 'p1': 2},
      )?.id,
      'p1',
    );
    expect(
      RideTogetherPolicy.pickGroupForRide(
        rideRouteId: 'saved-hd',
        catalogTourId: 'r-bodensee-road',
        groups: [session, planned],
        memberCounts: {'s1': 2, 'p1': 2},
      )?.id,
      'p1',
    );
  });

  test('Session-Gruppe nie als Treffen-Pin', () {
    final g = _g(
      id: 's1',
      route: RideTogetherPolicy.routeId,
      created: DateTime.utc(2026, 8, 19, 10),
    );
    expect(
      RideGroupPolicy.canShowMeetingOnExplore(
        g,
        isMember: true,
        now: DateTime.utc(2026, 8, 19, 11),
      ),
      isFalse,
    );
  });

  test('Look-JSON: Nähe ohne Koordinaten', () {
    final snap = RideTogetherCloud.parseLook(
      200,
      '{"me":"u1","joinCode":"AB12CD","nearby":[{"userId":"u2","label":"Sam","bucket":"beside"}],"inbound":[]}',
    );
    expect(snap.joinCode, 'AB12CD');
    expect(snap.nearby.single.userId, 'u2');
    expect(snap.nearby.single.bucket, 'beside');
  });

  test('Look-JSON: Outbound-Status', () {
    final snap = RideTogetherCloud.parseLook(
      200,
      '{"me":"u1","joinCode":"AB12CD","nearby":[],"inbound":[],'
      '"outbound":[{"id":"r1","toUserId":"u2","label":"Sam","status":"pending"}]}',
    );
    expect(snap.outbound.single.status, 'pending');
    expect(snap.activeOutbound?.toUserId, 'u2');
  });

  test('Chip-Kind: Code, Solo, Anfrage', () {
    expect(
      togetherChipKind(looking: true, joinCode: 'ABCDEF'),
      TogetherChipKind.code,
    );
    expect(
      togetherChipKind(looking: true),
      TogetherChipKind.soloWait,
    );
    expect(
      togetherChipKind(looking: true, outboundStatus: 'pending'),
      TogetherChipKind.wait,
    );
    expect(
      togetherChipKind(looking: false, inboundCount: 1),
      TogetherChipKind.inbound,
    );
    expect(togetherChipKind(looking: false), TogetherChipKind.idle);
  });
}

File: src/lib/community/rideTogether.test.ts
/**
 * npx tsx src/lib/community/rideTogether.test.ts
 */
import assert from "node:assert/strict";
import { canShowMeetingOnExplore } from "./rideGroup";
import {
  canAddSessionMember,
  canJoinSessionByCode,
  isFreerideRide,
  isSessionRouteId,
  nearbyFromLooks,
  pickRequestSession,
  sessionClosesAfterLeave,
  stopLookClosesSession,
  stopLookClosesSoloSession,
  sessionListedPublicly,
  listedPlannedGroups,
  matePair,
  pickGroupForRide,
  plannedMeetingOnly,
  sanitizeTogetherLabel,
  togetherBucket,
  RIDE_TOGETHER_ROUTE_ID,
} from "./rideTogether";

assert.equal(isSessionRouteId(RIDE_TOGETHER_ROUTE_ID), true);
assert.equal(isSessionRouteId("tour"), false);
assert.equal(isFreerideRide(null), true);
assert.equal(isFreerideRide("engine-1"), false);
assert.equal(canJoinSessionByCode("freeride"), true);
assert.equal(canJoinSessionByCode("tour"), false);
assert.equal(canAddSessionMember(19), true);
assert.equal(canAddSessionMember(20), false);
assert.equal(sessionClosesAfterLeave(1), false);
assert.equal(sessionClosesAfterLeave(0), true);
assert.equal(stopLookClosesSession(), false);
assert.equal(stopLookClosesSoloSession(), true);
assert.equal(sessionListedPublicly(), false);
assert.equal(pickRequestSession(1, 1), "from");
assert.equal(pickRequestSession(8, 1), "from");
assert.equal(pickRequestSession(1, 8), "to");
assert.equal(pickRequestSession(5, 12), "none");
assert.equal(pickRequestSession(20, 1), "from");

assert.deepEqual(
  nearbyFromLooks({
    selfLat: 49.4094,
    selfLng: 8.6948,
    selfId: "me",
    looks: [
      { user_id: "seeker", lat: 49.4094, lng: 8.6948, display_label: "Sam" },
      { user_id: "mate", lat: 49.4094, lng: 8.6948, display_label: "Hidden" },
      { user_id: "me", lat: 49.4094, lng: 8.6948, display_label: "Self" },
    ],
    labels: new Map(),
    excludeUserIds: ["mate"],
  }),
  [{ userId: "seeker", label: "Sam", bucket: "beside" }]
);

assert.equal(sanitizeTogetherLabel("  Luka   Flow  "), "Luka Flow");
assert.equal(sanitizeTogetherLabel("x".repeat(40)).length, 24);

assert.deepEqual(matePair("b", "a"), { lo: "a", hi: "b" });
assert.equal(matePair("a", "a"), null);

assert.equal(
  togetherBucket(49.4094, 8.6948, 49.4094, 8.6948),
  "beside"
);
assert.equal(
  togetherBucket(49.4094, 8.6948, 49.5, 8.8),
  null
);

const session = {
  id: "s1",
  savedRouteId: "freeride",
  createdAt: "2026-08-19T10:00:00.000Z",
  onServer: true,
};
const planned = {
  id: "p1",
  savedRouteId: "r-bodensee-road",
  catalogTourId: "r-bodensee-road",
  createdAt: "2026-08-19T09:00:00.000Z",
  onServer: true,
};

assert.equal(
  pickGroupForRide({
    rideRouteId: null,
    groups: [session, planned],
    memberCounts: { s1: 1, p1: 2 },
    idOf: (g) => g.id,
  }),
  null
);
assert.equal(
  pickGroupForRide({
    rideRouteId: null,
    groups: [session, planned],
    memberCounts: { s1: 2, p1: 2 },
    idOf: (g) => g.id,
  })?.id,
  "s1"
);
assert.equal(
  pickGroupForRide({
    rideRouteId: null,
    groups: [session],
    memberCounts: { s1: 20 },
    idOf: (g) => g.id,
  })?.id,
  "s1"
);
assert.equal(
  pickGroupForRide({
    rideRouteId: "r-bodensee-road",
    groups: [session, planned],
    memberCounts: { s1: 2, p1: 2 },
    idOf: (g) => g.id,
  })?.id,
  "p1"
);
assert.equal(
  pickGroupForRide({
    rideRouteId: "saved-hd",
    catalogTourId: "r-bodensee-road",
    groups: [session, planned],
    memberCounts: { s1: 2, p1: 2 },
    idOf: (g) => g.id,
  })?.id,
  "p1"
);
assert.equal(
  pickGroupForRide({
    rideRouteId: "r-other-city",
    groups: [session, planned],
    memberCounts: { s1: 2, p1: 2 },
    idOf: (g) => g.id,
  }),
  null
);

assert.deepEqual(
  listedPlannedGroups([session, planned]).map((g) => g.id),
  ["p1"]
);
assert.equal(
  plannedMeetingOnly(
    [
      {
        status: "open",
        startWindowStart: "2026-08-19T08:00:00.000Z",
        startWindowEnd: "2026-08-19T16:00:00.000Z",
        savedRouteId: "freeride",
      },
    ],
    new Date("2026-08-19T10:00:00.000Z")
  ),
  null
);

assert.equal(
  canShowMeetingOnExplore({
    visibility: "public",
    status: "open",
    startWindowEnd: "2026-08-19T16:00:00.000Z",
    isMember: false,
    now: new Date("2026-08-19T10:00:00.000Z"),
    savedRouteId: "freeride",
  }),
  false
);

console.log("rideTogether.test.ts ok");
